sktree/tree/_honest_tree.py

```diff
@@ -0,0 +1,338 @@
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix} of shape (n_samples, n_features)
+            The training input samples. Internally, it will be converted to
+            ``dtype=np.float32`` and if a sparse matrix is provided
+            to a sparse ``csc_matrix``.
+
+        y : array-like of shape (n_samples,) or (n_samples, n_outputs)
+            The target values (class labels) as integers or strings.
+
+        sample_weight : array-like of shape (n_samples,), default=None
+            Sample weights. If None, then samples are equally weighted. Splits
+            that would create child nodes with net zero or negative weight are
+            ignored while searching for a split in each node. Splits are also
+            ignored if they would result in any single class carrying a
+            negative weight in either child node.
+
+        check_input : bool, default=True
+            Allow to bypass several input checking.
+            Don't use this parameter unless you know what you do.
+
+        Returns
+        -------
+        self : HonestTreeClassifier
+            Fitted estimator.
+        """
+        if check_input:
+            X, y = check_X_y(X, y)
+        self._inherit_estimator_attributes()
+
+        # Account for bootstrapping too
+        if sample_weight is None:
+            sample_weight = np.ones((X.shape[0],), dtype=np.float64)
+
+        nonzero_indices = np.where(sample_weight > 0)[0]
+
+        self.structure_indices_ = np.random.choice(
+            nonzero_indices,
+            int((1 - self.honest_fraction) * len(nonzero_indices)),
+            replace=False,
+        )
+        self.honest_indices_ = np.setdiff1d(nonzero_indices, self.structure_indices_)
+
+        sample_weight[self.honest_indices_] = 0
+
+        # Learn structure on subsample
+        self.estimator_.fit(
+            X,
+            y,
+            sample_weight=sample_weight,
+            check_input=check_input,
+        )
+
+        if self.n_outputs_ > 1:
+            raise NotImplementedError(
+                "Multi-target honest trees not yet \
+                implemented"
+            )
+
+        # update the number of classes, unsplit
+        # self.n_samples_, self.n_features_in_ = X.shape
+        if y.ndim == 1:
+            # reshape is necessary to preserve the data contiguity against vs
+            # [:, np.newaxis] that does not.
+            y = np.reshape(y, (-1, 1))
+        check_classification_targets(y)
+        y = np.copy(y).astype(int)
+        # Normally called by super
+        X = self.estimator_._validate_X_predict(X, True)
+        # Fit leaves using other subsample
+        honest_leaves = self.tree_.apply(X[self.honest_indices_])
+
+        self.tree_.value[:, :, :] = 0
+        for leaf_id, yval in zip(honest_leaves, y[self.honest_indices_, 0]):
+            self.tree_.value[leaf_id][0, yval] += 1
```

Comment:
This can be migrated to a private function perhaps called `_set_leaf_nodes(self, X, y)`, where (X, y) are a data pair where X is used to traverse the already built tree and y is used to set the leaf nodes. 

This can then lead us to easily supporting `PropensityTree`